model/src/com/redhat/ceylon/model/typechecker/model/TypeDeclaration.java

```diff
@@ -996,14 +996,15 @@ private void mergeMembers(
      *         the given list of types
      */
     public boolean isDisjoint(TypeDeclaration td) {
-        if (this instanceof ClassOrInterface &&
-            td instanceof ClassOrInterface &&
-                equals(td)) {
+        if (!(this instanceof ClassOrInterface)
+                || !(td instanceof ClassOrInterface)) {
             return false;
         }
-        if (this instanceof TypeParameter &&
-                td instanceof TypeParameter &&
-                    equals(td)) {
+        if (this instanceof Class
+                && td instanceof Class) {
+            return false;
```

Comment:
This method returns false in a lot of cases where the declarations are disjoint. This particular one is covered by `ModelUtil.emptyMeet()`, I believe:

```
        if (p.isClass() && q.isClass() ||
            p.isInterface() && q.isNull() ||
            q.isInterface() && p.isNull()) {
            if (!qd.inherits(pd) &&
                !pd.inherits(qd)) {
                return true;
            }
        }
```